src/domain/datasets/services/src/services/statistics/dataset_statistics_update_handler.rs

```diff
@@ -0,0 +1,145 @@
+
+impl DatasetStatisticsUpdateHandler {
+    async fn update_dataset_statistics(
+        &self,
+        dataset_stats_repo: &dyn DatasetStatisticsRepository,
+        target: ResolvedDataset,
+        updated_message: &DatasetReferenceMessageUpdated,
+    ) -> Result<(), InternalError> {
+        let increment = compute_dataset_statistics_increment(
+            target.as_metadata_chain(),
+            &updated_message.new_block_hash,
+            updated_message.maybe_prev_block_hash.as_ref(),
+        )
+        .await?;
+
+        // If we have seen Seed, the dataset has diverged,
+        // and the increment represents the updated situation
+        if increment.seen_seed {
+            dataset_stats_repo
+                .set_dataset_statistics(
+                    &updated_message.dataset_id,
+                    &updated_message.block_ref,
+                    increment.statistics,
+                )
+                .await
+                .int_err()
+        } else {
+            // Otheriwse, load previous stats
+            let older_stats = match dataset_stats_repo
+                .get_dataset_statistics(&updated_message.dataset_id, &updated_message.block_ref)
+                .await
+            {
+                Ok(stats) => Ok(stats),
+                Err(GetDatasetStatisticsError::NotFound(_)) => Ok(DatasetStatistics::default()),
+                Err(GetDatasetStatisticsError::Internal(e)) => Err(e),
+            }?;
+
+            // Save old + increment as new stats
+            let new_stats = older_stats + increment.statistics;
+            dataset_stats_repo
+                .set_dataset_statistics(
+                    &updated_message.dataset_id,
+                    &updated_message.block_ref,
+                    new_stats,
+                )
+                .await
+                .int_err()
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+impl MessageConsumer for DatasetStatisticsUpdateHandler {}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl MessageConsumerT<DatasetReferenceMessage> for DatasetStatisticsUpdateHandler {
+    #[tracing::instrument(
+        level = "debug",
+        skip_all,
+        name = "DatasetStatisticsUpdateHandler[DatasetReferenceMessage]"
+    )]
+    async fn consume_message(
+        &self,
+        transaction_catalog: &Catalog,
+        message: &DatasetReferenceMessage,
+    ) -> Result<(), InternalError> {
+        tracing::debug!(received_message = ?message, "Received dataset reference message");
+
+        match message {
+            DatasetReferenceMessage::Updated(updated_message) => {
+                let dataset_registry = transaction_catalog
+                    .get_one::<dyn DatasetRegistry>()
+                    .unwrap();
```

Comment:
Why are we passing catalog as a parameter and pulling things out manually?
Why can't `DatasetStatisticsUpdateHandler` inject its dependencies?
Also why this stateless type needs to be `Singleton`?

Reply:
`Singleton` - good catch, copy pasted unnoticed.
I've found 2 more listeners with a similar problem.